Fix soft_delete,User has been deleted instance,

Farice4 opened the pull request:
The user quota have never changed.
Redmine #7102 http://192.168.15.2/issues/7102
Signed-off-by: fabian cybing4@gmail.com

Fix soft_delete,abnormal quota usage after restore by admin
Use upstream commit: c474b44bdcdda223b0ec90b7fa1673501cf690e7
Redmine #7099 http://192.168.15.2/issues/7099

ankit-agrawal committed:
'deleted' filter does not work properly

While calling server list with deleted filter set to true/false, it
always returns deleted instances to admin, because the code didn't
convert string to boolean value. So even if we pass deleted='false' or
any string value, it always returns True. Converted 'deleted' filter
from string to its boolean value and set default to False if called with
an invalid string.

Use upstream commit: 5c7c852dc0e4ac44e167585938a2b944c1156d10

Signed-off-by: fabian <cybing4@gmail.com>

apporc committed:
fix soft-delete for 1.0.1

Signed-off-by: apporc <appleorchard2000@gmail.com>

blkart commented on nova/compute/api.py:
```diff
@@ -1773,10 +1774,20 @@ def soft_delete(self, context, instance):
+            raise exception.InstanceTerminationFailure(reason=reason)
+        elif instance['vm_state'] == vm_states.ACTIVE or \
+                instance['vm_state'] == vm_states.STOPPED:
+            self._delete(context, instance, 'soft_delete',
```
只允许删除 ACTIVE 和 STOPPED 状态的虚拟机是否妥当

Farice4 replied:
代码中已经全部放开软删除操作．代码已提交．

2016-06-24 16:25 GMT+08:00 blkart notifications@github.com:

> In nova/compute/api.py
> https://github.com/eayunstack/nova/pull/37#discussion_r68367808:
> 
> > @@ -1773,10 +1774,20 @@ def soft_delete(self, context, instance):
> >          """Terminate an instance."""
> >          LOG.debug('Going to try to soft delete instance',
> > 
> > ##                    instance=instance)
> > -        self._delete(context, instance, 'soft_delete', self._do_soft_delete,
> > -                     task_state=task_states.SOFT_DELETING,
> > -                     deleted_at=timeutils.utcnow())
> > -        if instance['vm_state'] == vm_states.SOFT_DELETED:
> > -            LOG.warn("%s is soft delete state" % (instance.uuid))
> > -            reason = ("The instance is %s state" % (instance.vm_state))
> > -            raise exception.InstanceTerminationFailure(reason=reason)
> > -        elif instance['vm_state'] == vm_states.ACTIVE or \
> > -                instance['vm_state'] == vm_states.STOPPED:
> > -            self._delete(context, instance, 'soft_delete',
> 
> 只允许删除 ACTIVE 和 STOPPED 状态的虚拟机是否妥当
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/eayunstack/nova/pull/37/files/77d779d01ef86664194b0ef8116f62792a28ceb3#r68367808,
> or mute the thread
> https://github.com/notifications/unsubscribe/AKMiZkqEkVaItCL83xD3eBxLM3EHMpkAks5qO5R-gaJpZM4I0Frh
> .

blkart commented on nova/compute/api.py:
```diff
@@ -1799,8 +1810,9 @@ def restore(self, context, instance):
         num_instances, quotas = self._check_num_instances_quota(
-                context, flavor, 1, 1)
+            context, flavor, 1, 1, project_id=project_id, user_id=user_id)
 
```
一个 commit 里包含了解决两个不同问题的代码

Farice4 committed:
Fix soft_delete,abnormal quota usage after restore by admin
Use upstream commit: c474b44bdcdda223b0ec90b7fa1673501cf690e7
Redmine #7099 http://192.168.15.2/issues/7099

Signed-off-by: fabian <cybing4@gmail.com>

Farice4 committed:
Fix soft_delete,User has been deleted instance,
The user quota have never changed.
Redmine #7102 http://192.168.15.2/issues/7102

Signed-off-by: fabian <cybing4@gmail.com>

blkart commented on nova/tests/compute/test_compute_api.py:
```diff
@@ -2066,8 +2066,31 @@ def test_check_injected_file_quota_onset_file_content_limit(self):
+
+    @mock.patch('nova.objects.Quotas.commit')
+    @mock.patch('nova.objects.Quotas.reserve')
+    @mock.patch('nova.objects.Quotas.save')
```
上游是 @mock.patch('nova.objects.Instance.save')

Farice4 replied:
这个地方已修改，并已经提交

blkart commented on nova/compute/api.py:
```diff
@@ -1774,10 +1774,15 @@ def soft_delete(self, context, instance):
-                     task_state=task_states.SOFT_DELETING,
-                     deleted_at=timeutils.utcnow())
+        if instance['vm_state'] == vm_states.SOFT_DELETED:
+            LOG.warn("%s is soft delete state" % (instance.uuid))
```
日志信息改成“Instance %s is already in soft-delete state,  ignoring this request."是不是更好些。

blkart commented on nova/compute/api.py:
```diff
@@ -1774,10 +1774,15 @@ def soft_delete(self, context, instance):
+        if instance['vm_state'] == vm_states.SOFT_DELETED:
+            LOG.warn("%s is soft delete state" % (instance.uuid))
+            reason = ("The instance is %s state" % (instance.vm_state))
+            raise exception.InstanceTerminationFailure(reason=reason)
```
感觉这里不需要抛异常

blkart commented on nova/tests/api/openstack/compute/plugins/v3/test_servers.py:
```diff
@@ -1065,6 +1065,54 @@ def fake_get_all(compute_self, context, search_opts=None,
+    @mock.patch.object(compute_api.API, 'get_all')
+    def test_get_servers_deleted_filter_invalid_str(self, mock_get_all):
+        server_uuid = str(uuid.uuid4())
+        db_list = [fakes.stub_instance(100, uuid=server_uuid)]
```
上游 commit 这里有空格

blkart commented on nova/tests/api/openstack/compute/plugins/v3/test_servers.py:
```diff
@@ -1065,6 +1065,54 @@ def fake_get_all(compute_self, context, search_opts=None,
+
+    @mock.patch.object(compute_api.API, 'get_all')
+    def test_get_servers_deleted_filter_invalid_str(self, mock_get_all):
+        server_uuid = str(uuid.uuid4())
```
上游 commit 这里有空行

Farice4 replied:
已修改，代码已提交．

blkart commented:
- 6a17f00074801073e41cf9f5a489ac281a80b79b 这个 commit 的 commit 信息需要修改一下，这个 commit 并不是对 “nova list 未单独显示 SOFT_DELETED 状态的虚拟机“ 问题的修复。

pengdake commented on nova/compute/api.py:
```diff
@@ -362,7 +362,8 @@ def _check_num_instances_quota(self, context, instance_type, min_count,
         try:
             quotas = objects.Quotas(context)
             quotas.reserve(context, instances=max_count,
-                           cores=req_cores, ram=req_ram)
```
这里上游没有带context参数

Farice4 replied:
这个地方上游没有带context，是因为上游在修复这个bug之前已经修复了多个提交，其中一个提交中就是将quotas.reserve(context,
instances=max_count 改为quotas.reserve( instances=max_coun
去掉了context，见（4031272fe93d1046cff4c0cc788c8e78db944419）
在修复该bug的目的是admin
做restore没有带project_id与user_id，因此恢复默认使用了admin来进行恢复，因此在不做大的改动原则下，添加了project_id,
user_id ，从目前测试来看没有发现问题．

2016-07-08 11:08 GMT+08:00 daemon notifications@github.com:

> In nova/compute/api.py
> https://github.com/eayunstack/nova/pull/37#discussion_r70020387:
> 
> > @@ -362,7 +362,8 @@ def _check_num_instances_quota(self, context, instance_type, min_count,
> >          try:
> >              quotas = objects.Quotas(context)
> >              quotas.reserve(context, instances=max_count,
> > -                           cores=req_cores, ram=req_ram)
> 
> 这里上游没有带context参数
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/eayunstack/nova/pull/37/files/b9f59ddec6dfb4785ef13559f31c37d1239b21e7#r70020387,
> or mute the thread
> https://github.com/notifications/unsubscribe/AKMiZlsZ4KvSLVGnDuGxCWpzLS1daYTRks5qTb8UgaJpZM4I0Frh
> .

pengdake commented on nova/compute/api.py:
```diff
@@ -1773,10 +1774,15 @@ def soft_delete(self, context, instance):
+                         self._do_soft_delete,
+                         task_state=task_states.SOFT_DELETING,
+                         deleted_at=timeutils.utcnow())
 
```
我觉得既然是关于配额处理的问题，最好放到_delete里面，与其他关于配额处理的代码放一块。可以在https://git.openstack.org/cgit/openstack/nova/commit/?id=4546ce0012b526cd370533ba5a3f2f4107d249a2的 基础上加以修改

Farice4 replied:
这里不只是配额的问题，还涉及到测试时，如果第一次删除没有提示信息，你是否认为已经成功了？
基于uuid执行第二次删除依旧没有提示信息，你是否认为还是成功了？ 那你这样一直执行下去，你是否认为一直都成功了？你认为这是否是一个BUG ?
这个地方之所以不用上游的删除后，在进行配额回滚，是从根本上就不允许，已经处于软删除状态的云主机再次执行软删除操作．

2016-07-08 11:51 GMT+08:00 daemon notifications@github.com:

> In nova/compute/api.py
> https://github.com/eayunstack/nova/pull/37#discussion_r70022488:
> 
> > @@ -1773,10 +1774,15 @@ def soft_delete(self, context, instance):
> >          """Terminate an instance."""
> >          LOG.debug('Going to try to soft delete instance',
> > 
> > ##                    instance=instance)
> > -        self._delete(context, instance, 'soft_delete', self._do_soft_delete,
> > -                     task_state=task_states.SOFT_DELETING,
> > -                     deleted_at=timeutils.utcnow())
> > -        if instance['vm_state'] == vm_states.SOFT_DELETED:
> > -            LOG.warn("Instance %s is soft delete state" % (instance.uuid))
> > -            reason = ("The instance is %s state" % (instance.vm_state))
> > -            raise exception.InstanceTerminationFailure(reason=reason)
> > -        else:
> > -            self._delete(context, instance, 'soft_delete',
> > -                         self._do_soft_delete,
> > -                         task_state=task_states.SOFT_DELETING,
> > -                         deleted_at=timeutils.utcnow())
> 
> 我觉得既然是关于配额处理的问题，最好放到_delete里面，与其他关于配额处理的代码放一块。可以在
> https://git.openstack.org/cgit/openstack/nova/commit/?id=4546ce0012b526cd370533ba5a3f2f4107d249a2的
> 基础上加以修改
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/eayunstack/nova/pull/37/files/b9f59ddec6dfb4785ef13559f31c37d1239b21e7#r70022488,
> or mute the thread
> https://github.com/notifications/unsubscribe/AKMiZudlsCaMfcjfycbWnoKed1zd1nipks5qTclJgaJpZM4I0Frh
> .

pengdake replied:
我不反对打印提示信息，我只是建议在那基础上修改，当然这样也没什么问题

Farice4 replied:
这个地方引入异常抛出目的在于告诉测试以及后台使用人员，当一个处于已经删除状态的云主机，再次进行删除时会提示错误，如果不提示错误，测试或者后台使用人员会觉得一直进行删除，但都没有报不存在或其它错误，会感觉这是一个bug，怎么删除都没有问题的bug
．

2016-07-08 13:35 GMT+08:00 daemon notifications@github.com:

> In nova/compute/api.py
> https://github.com/eayunstack/nova/pull/37#discussion_r70027047:
> 
> > @@ -1773,10 +1774,15 @@ def soft_delete(self, context, instance):
> >          """Terminate an instance."""
> >          LOG.debug('Going to try to soft delete instance',
> > 
> > ##                    instance=instance)
> > -        self._delete(context, instance, 'soft_delete', self._do_soft_delete,
> > -                     task_state=task_states.SOFT_DELETING,
> > -                     deleted_at=timeutils.utcnow())
> > -        if instance['vm_state'] == vm_states.SOFT_DELETED:
> > -            LOG.warn("Instance %s is soft delete state" % (instance.uuid))
> > -            reason = ("The instance is %s state" % (instance.vm_state))
> > -            raise exception.InstanceTerminationFailure(reason=reason)
> > -        else:
> > -            self._delete(context, instance, 'soft_delete',
> > -                         self._do_soft_delete,
> > -                         task_state=task_states.SOFT_DELETING,
> > -                         deleted_at=timeutils.utcnow())
> 
> 我不反对打印提示信息，我只是建议在那基础上修改，当然这样也没什么问题
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/eayunstack/nova/pull/37/files/b9f59ddec6dfb4785ef13559f31c37d1239b21e7#r70027047,
> or mute the thread
> https://github.com/notifications/unsubscribe/AKMiZkO1CCGquwcwwzbYweDkk-86i13rks5qTeGFgaJpZM4I0Frh
> .

Farice4 commented:
这个commit提交信息已于７月７日进行修改，修改为上游相同的commit信息，最初测试时，因为其它测试与soft-delete测试交叉，修改了soft-delete保留周期为２分钟左右，因此在删除后，错过了查询soft-delete状态云主机，导致云主机状态变成了delete状态．

2016-07-05 16:52 GMT+08:00 blkart notifications@github.com:

> - 6a17f00
>   https://github.com/eayunstack/nova/commit/6a17f00074801073e41cf9f5a489ac281a80b79b
>   这个 commit 的 commit 信息需要修改以下，这个 commit 并不是对 “nova list 未单独显示 SOFT_DELETED
>   状态的虚拟机“ 问题的修复。
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/eayunstack/nova/pull/37#issuecomment-230422493, or mute
> the thread
> https://github.com/notifications/unsubscribe/AKMiZosY0JxY80JbgAnkgWS11fOsY6E1ks5qShtigaJpZM4I0Frh
> .
